frontendUiTableImpl/src/main/java/com/virtuslab/gitmachete/frontend/ui/impl/table/EnhancedGraphTable.java

```diff
@@ -216,67 +213,57 @@ private void doAutomaticDiscover(Path macheteFilePath) {
-            .filterTry(repositorySnapshot -> repositorySnapshot.getRootBranches().size() > 0,
-                __ -> new EmptyRepositoryException())
-            .map(repositorySnapshot -> {
-              var branchLayoutWriter = project.getService(BranchLayoutWriterProvider.class).getBranchLayoutWriter();
-              var branchLayout = repositorySnapshot.getBranchLayout().getOrNull();
-              return DiscoverResult.of(branchLayoutWriter, branchLayout, repositorySnapshot);
-            })
-            .filterTry(result -> result.branchLayout != null, __ -> new EmptyBranchLayoutException())
-            .andThenTry(
-                result -> result.branchLayoutWriter.write(macheteFilePath, result.branchLayout, /* backupBranchLayout */ true));
-      }
-      // This method is invoked on UI dispatch thread
-      @UIEffect
-      @Override
-      public void onSuccess() {
-        discoverResult.onFailure(exception -> {
-          Class<? extends Throwable> exceptionClass = exception.getClass();
-          if (EmptyRepositoryException.class.equals(exceptionClass)) {
-            setTextForEmptyTable(format(getString("string.GitMachete.EnhancedGraphTable.empty-machete-file.upper-text"),
-                macheteFilePath.toString()));
-          } else if (EmptyBranchLayoutException.class.equals(exceptionClass)) {
-            LOG.error("Can't get branch layout from repository snapshot");
-          } else {
-            VcsNotifier.getInstance(project).notifyError(
+        var discoverRunResult = Try.of(() -> RuntimeBinding.instantiateSoleImplementingClass(IGitMacheteRepositoryCache.class)
+            .getInstance(mainDirPath, gitDirPath).discoverLayoutAndCreateSnapshot());
+
+        if (!discoverRunResult.isSuccess()) {
+          var exception = discoverRunResult.getCause();
+          GuiUtils.invokeLaterIfNeeded(() -> VcsNotifier.getInstance(project).notifyError(
+              getString("string.GitMachete.EnhancedGraphTable.automatic-discover.cant-discover-layout-error-title"),
+              exception.getMessage() != null ? exception.getMessage() : ""), NON_MODAL);
+          return;
+        }
+
+        var repositorySnapshot = discoverRunResult.get();
+
+        if (repositorySnapshot.getRootBranches().size() == 0) {
+          GuiUtils.invokeLaterIfNeeded(
+              () -> setTextForEmptyTable(format(getString("string.GitMachete.EnhancedGraphTable.empty-machete-file.upper-text"),
+                  macheteFilePath.toString())),
+              NON_MODAL);
+          return;
+        }
+
+        var branchLayoutWriter = project.getService(BranchLayoutWriterProvider.class).getBranchLayoutWriter();
+        var branchLayout = repositorySnapshot.getBranchLayout().getOrNull();
+
+        if (branchLayout == null) {
+          LOG.error("Can't get branch layout from repository snapshot");
+          return;
+        }
+
+        @SuppressWarnings("guieffect") // To allow lambda invocation in `onSuccess` method
```

Comment:
Hmm I'm afraid that the GuiEffect error here is justified :thinking:
See the comment below to `setModel`

Reply:
Done, reworked